### PR TITLE
Revamp exceptions thrown by the hub client - part 2

### PR DESCRIPTION
**Pull request by brycewang-microsoft**
+ simplify ctors in `IotHubClientException`.
+ rename `IotHubStatusCode` to `IotHubErrorCode`, which aligns to what we have for the hub **service**.
+ add IsTransient check into e2e tests and unit tests
+ remove `AmqpIotResourceException`

### CODE REVIEW

File: iothub/device/src/Exceptions/ExceptionHandlingHelper.cs [hunk 1]
```diff
@@ -15,16 +15,66 @@ internal class ExceptionHandlingHelper
...
-                { HttpStatusCode.PreconditionFailed, async (response) => new IotHubClientException(await GetExceptionMessageAsync(response).ConfigureAwait(false), isTransient: false, IotHubStatusCode.DeviceMessageLockLost) },
-                { HttpStatusCode.RequestEntityTooLarge, async (response) => new IotHubClientException(await GetExceptionMessageAsync(response).ConfigureAwait(false), isTransient: false, IotHubStatusCode.MessageTooLarge) },
-                { HttpStatusCode.InternalServerError, async (response) => new IotHubClientException(await GetExceptionMessageAsync(response).ConfigureAwait(false), isTransient: true, IotHubStatusCode.ServerError) },
-                { HttpStatusCode.ServiceUnavailable, async (response) => new IotHubClientException(await GetExceptionMessageAsync(response).ConfigureAwait(false), innerException: null, isTransient: true, IotHubStatusCode.ServerBusy) },
-                { (HttpStatusCode)429, async (response) => new IotHubClientException(await GetExceptionMessageAsync(response).ConfigureAwait(false), innerException: null, isTransient: true, IotHubStatusCode.Throttled) }
+                {
```
Comment (brycewang-microsoft) [inline iothub/device/src/Exceptions/ExceptionHandlingHelper.cs:18]: formatting this `mappings` as each line was too long

File: iothub/device/src/Exceptions/IotHubClientException.cs [hunk 2]
```diff
@@ -211,5 +176,31 @@ public override void GetObjectData(SerializationInfo info, StreamingContext cont
             info.AddValue(IsTransientValueSerializationStoreName, IsTransient);
             info.AddValue(TrackingIdValueSerializationStoreName, TrackingId);
         }
+
+        private Dictionary<IotHubErrorCode, bool> mapping = new Dictionary<IotHubErrorCode, bool>
```
Comment (brycewang-microsoft) [inline iothub/device/src/Exceptions/IotHubClientException.cs:180]: add a mapping between the error code and IsTrasient flag

File: e2e/test/iothub/twin/TwinE2ETests.cs [hunk 3]
```diff
@@ -687,7 +687,7 @@ public async Task Twin_Client_SetETag_Works(IotHubClientTransportProtocol transp
             Func<Task> act = async () => { twin = await _serviceClient.Twins.UpdateAsync(testDevice.Id, twin, true).ConfigureAwait(false); };
             var error = await act.Should().ThrowAsync<IotHubServiceException>("Expected test to throw a precondition failed exception since it updated a twin with an out of date ETag");
             error.And.StatusCode.Should().Be(HttpStatusCode.PreconditionFailed);
-            error.And.ErrorCode.Should().Be(IotHubErrorCode.PreconditionFailed);
+            error.And.ErrorCode.Should().Be(Common.Exceptions.IotHubErrorCode.PreconditionFailed);
```
Comment (drwill-ms) [inline e2e/test/iothub/twin/TwinE2ETests.cs:690]: Can we fix the namespace here? It should just be in the project root.

File: iothub/device/src/Exceptions/ExceptionHandlingHelper.cs [hunk 4]
```diff
@@ -15,16 +15,66 @@ internal class ExceptionHandlingHelper
...
-                { HttpStatusCode.RequestEntityTooLarge, async (response) => new IotHubClientException(await GetExceptionMessageAsync(response).ConfigureAwait(false), isTransient: false, IotHubStatusCode.MessageTooLarge) },
-                { HttpStatusCode.InternalServerError, async (response) => new IotHubClientException(await GetExceptionMessageAsync(response).ConfigureAwait(false), isTransient: true, IotHubStatusCode.ServerError) },
-                { HttpStatusCode.ServiceUnavailable, async (response) => new IotHubClientException(await GetExceptionMessageAsync(response).ConfigureAwait(false), innerException: null, isTransient: true, IotHubStatusCode.ServerBusy) },
-                { (HttpStatusCode)429, async (response) => new IotHubClientException(await GetExceptionMessageAsync(response).ConfigureAwait(false), innerException: null, isTransient: true, IotHubStatusCode.Throttled) }
+                { 
+                    HttpStatusCode.NoContent, async (response) =>
```
Comment (drwill-ms) [inline iothub/device/src/Exceptions/ExceptionHandlingHelper.cs:19]: key and value on separate lines

File: iothub/device/src/Exceptions/IotHubClientException.cs (hunk 2 above)
Comment (drwill-ms) [reply iothub/device/src/Exceptions/IotHubClientException.cs:180]: Since it is a boolean, you could simply have a hashset of codes that are transient.

Comment (drwill-ms) [inline iothub/device/src/Exceptions/IotHubClientException.cs:180]: fields at the top of the class

File: iothub/device/src/Exceptions/IotHubErrorCode.cs [hunk 5]
```diff
@@ -6,7 +6,7 @@ namespace Microsoft.Azure.Devices.Client.Exceptions
     /// <summary>
     /// The IoT Hub status code.
```
Comment (drwill-ms) [inline iothub/device/src/Exceptions/IotHubErrorCode.cs:7]: Fix doc comment

File: iothub/device/src/Exceptions/IotHubErrorCode.cs [hunk 6]
```diff
@@ -6,7 +6,7 @@ namespace Microsoft.Azure.Devices.Client.Exceptions
...
     /// </summary>
-    public enum IotHubStatusCode
+    public enum IotHubErrorCode
```
Comment (drwill-ms) [inline iothub/device/src/Exceptions/IotHubErrorCode.cs:9]: I'd like us to avoid naming conflicts with the service SDK. Can we make these distinct somehow?

**Review by drwill-ms — APPROVED** [review]
Comments are inline; signing off optimistically.

**Pull request merged**